### ISSUE TITLE
How to mock navigation arguments for testing flutter screen widgets?

### ISSUE THREAD

**Issue by DFelten** [dart-lang/mockito#199]
I would like to write a widget test for a screen widget in flutter. The problem is, that this widget uses a variable from the navigation argument and I'm not sure how to mock this variable with mockito.

This is the example screen:

```dart
class TestScreen extends StatefulWidget {
  static final routeName = Strings.contact;

  @override
  _TestScreenState createState() => _TestScreenState();
}

class _TestScreenState extends State<TestScreen> {
  Contact _contact;

  @override
  Widget build(BuildContext context) {
    _contact = ModalRoute.of(context).settings.arguments;

    return Scaffold(
      appBar: AppBar(title: Text(Strings.contact)),
      body: Text(_contact.name),
    );
  }
}
```

With this command I open the screen

```dart
Navigator.pushNamed(context, TestScreen.routeName, arguments: contact);`
```

Normally I would mock some components, but I'm not sure how to mock the screen arguments. I hope it works something like this. However, I do not know what I can exactly mock.

```dart
when(screenArgument.fetchData(any))
    .thenAnswer((_) async => expectedContact);
```

This is the current test, which of course is not working since _contact is null:

```dart
void main() {
  testWidgets('contact fields should be filled with data from argument', (WidgetTester tester) async {
    // GIVEN
    final testScreen = TestApp(widget: TestScreen());

    // WHEN
    await tester.pumpWidget(testScreen);

    // THEN
    expect(find.text("test"), findsOneWidget);
  });
}
```
An ugly way would be to use constructor parameters for the screen only for testing, but I want to avoid that.

Maybe someone of you knows how to best test such screen widgets.

**Comment by srawlins** [comment]
I think your best shot is to ask this on Stack Overflow. https://stackoverflow.com/questions/tagged/flutter?sort=frequent

None of the Mockito maintainers are Flutter experts, so I doubt we can help you here. I've read your (quite detailed) example a few times and I just don't have a good context on how Flutter widget tests are typically written, what methods you explicitly call in a test, etc. The Stack Overflow community should have an answer for you.

**Comment by DFelten** [comment]
Thanks for the answer. I already tried to get help on [Stack Overflow](https://stackoverflow.com/questions/57005472/how-to-mock-navigation-arguments-for-testing-flutter-screen-widgets), but currently there is no answer after three days. So I thought the chances of getting help here would be better.

**Comment by flavioccf** [comment]
Hi @DFelten did you find any work arounds for this situation?

**Comment by DFelten** [comment]
When using the package [auto_route](https://pub.dev/packages/auto_route) you could define the parameters as screen arguments. Then testing becomes easier.

**Comment by davidobber** [comment]
I'm just starting with flutter so probably it's not the best solution, anyway this is what I've done today:

I've create a NavigatorScreen widget and I've added it to my test fixtures. Then I use it as Home just to tell it to call the screen I want to test, passing it some arguments. 

Then I've passed it to a FakeMyApp widget (actually it's more complex than this and maybe it's out of scope) to have the NavigatorScreen as my home screen.

calling _buildMainPage() and then await tester.pumpAndSettle() ensures that MyScreen() will be called, receiving the parameters as if it was called by a real page.

~~~dart
class NavigatorScreen extends StatelessWidget {
  Widget screenToGo;
  Object arguments;

  NavigatorScreen({this.screenToGo, this.arguments});

  @override
  Widget build(BuildContext context) {
    Future.delayed(Duration.zero).then((value) {
      Navigator.push(context, MaterialPageRoute(
          builder: (BuildContext context) => screenToGo,
          settings: RouteSettings(
              arguments: arguments
          )));
    });
    return Container();
  }
}


//simplified...
class FakeMyApp extends StatelessWidget {
  Widget homeScreen;

  FakeMyApp({this.homeScreen});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      routes: {
        "/": (_) => homeScreen,
      },
    );
  }
}

void main() {

  group('Test.... ', () {

   Future<void> _buildMainPage(WidgetTester tester) async {
      await tester.pumpWidget(FakeMyApp(
          homeScreen: NavigatorScreen(
             screenToGo: MyScreen(),
             arguments: { "id": "xyz" }
         )
      ));
    }

    testWidgets(
        'navigation test', (WidgetTester tester) async {

          await _buildMainPage(tester);
          await tester.pumpAndSettle();
          await expect(find.byType(MyScreen), findsOneWidget);
        });
  });

}
~~~
